webui/react/src/components/CompareHeatMaps.tsx

```diff
@@ -0,0 +1,362 @@
+
+  const { data } = metricData;
+
+  useEffect(() => {
+    if (activeHParam) {
+      galleryModal.open();
+    }
+  }, [activeHParam, galleryModal]);
+
+  const smallerIsBetter = useMemo(() => {
+    if (selectedMetric && selectedMetric.group === MetricType.Validation) {
+      const selectedExperimentsWithMetric = selectedExperiments.filter((exp) => {
+        return selectedMetric.name === exp?.experiment?.config?.searcher?.metric;
+      });
+
+      return selectedExperimentsWithMetric.some((exp) => {
+        return exp?.experiment?.config?.searcher?.smallerIsBetter;
+      });
+    }
+    return undefined;
+  }, [selectedMetric, selectedExperiments]);
+
+  const colorScale = useMemo(() => {
+    return getColorScale(ui.theme, chartData?.metricRange, smallerIsBetter);
+  }, [chartData, smallerIsBetter, ui.theme]);
+
+  const chartProps = useMemo(() => {
+    if (!chartData || !selectedMetric) return undefined;
+
+    const props: Record<string, UPlotScatterProps> = {};
+    const rgbaStroke0 = str2rgba(colorScale[0].color);
+    const rgbaStroke1 = str2rgba(colorScale[1].color);
+    const rgbaFill0 = structuredClone(rgbaStroke0);
+    const rgbaFill1 = structuredClone(rgbaStroke1);
+    rgbaFill0.a = 0.3;
+    rgbaFill1.a = 0.3;
+    const fill = [rgba2str(rgbaFill0), rgba2str(rgbaFill1)].join(' ');
+    const stroke = [rgba2str(rgbaStroke0), rgba2str(rgbaStroke1)].join(' ');
+
+    selectedHParams.forEach((hParam1) => {
+      selectedHParams.forEach((hParam2) => {
+        const key = generateHpKey(hParam1, hParam2);
+        const xLabel = hParam2;
+        const yLabel = hParam1;
+        const title = `${yLabel} (y) vs ${xLabel} (x)`;
+        const xHpLabels = chartData?.hpLabels[hParam2];
+        const yHpLabels = chartData?.hpLabels[hParam1];
+        const isXLogarithmic = chartData?.hpLogScales[hParam2];
+        const isYLogarithmic = chartData?.hpLogScales[hParam1];
+        const isXCategorical = xHpLabels?.length !== 0;
+        const isYCategorical = yHpLabels?.length !== 0;
+        const xScaleKey = isXCategorical ? 'xCategorical' : isXLogarithmic ? 'xLog' : 'x';
+        const yScaleKey = isYCategorical ? 'yCategorical' : isYLogarithmic ? 'yLog' : 'y';
+        const xSplits = isXCategorical
+          ? new Array(xHpLabels.length).fill(0).map((_x, i) => i)
+          : undefined;
+        const ySplits = isYCategorical
+          ? new Array(yHpLabels.length).fill(0).map((_x, i) => i)
+          : undefined;
+        const xValues = isXCategorical ? xHpLabels : undefined;
+        const yValues = isYCategorical ? yHpLabels : undefined;
+
+        props[key] = {
+          data: [
+            null,
+            [
+              chartData?.hpValues[hParam2] || [],
+              chartData?.hpValues[hParam1] || [],
+              null,
+              chartData?.hpMetrics[key] || [],
+              chartData?.hpMetrics[key] || [],
+              chartData?.trialIds || [],
+            ],
+          ] as FacetedData,
```

Comment:
i think we should avoid `as` as much as possible cuz it leads to the bug easily

Reply:
discussed with caleb and I think it makes sense to keep this particular assertion for now

Reply:
nvm, did remove